packages/cli/src/decorators/controller.registry.ts

```diff
@@ -0,0 +1,136 @@
+import { Container, Service } from 'typedi';
+import { Router } from 'express';
+import type { Application, Request, Response, RequestHandler } from 'express';
+import { rateLimit as expressRateLimit } from 'express-rate-limit';
+
+import { AuthService } from '@/auth/auth.service';
+import config from '@/config';
+import { UnauthenticatedError } from '@/errors/response-errors/unauthenticated.error';
+import { inProduction, RESPONSE_ERROR_MESSAGES } from '@/constants';
+import type { BooleanLicenseFeature } from '@/Interfaces';
+import { License } from '@/License';
+import type { AuthenticatedRequest } from '@/requests';
+import { send } from '@/ResponseHelper'; // TODO: move `ResponseHelper.send` to this file
+import { userHasScope } from '@/permissions/checkAccess';
+
+import type {
+	AccessScope,
+	Controller,
+	ControllerMetadata,
+	HandlerName,
+	RateLimit,
+	RouteMetadata,
+} from './types';
+
+const registry = new Map<Function, ControllerMetadata>();
+
+export const getControllerMetadata = (controllerClass: Function) => {
+	let metadata = registry.get(controllerClass);
+	if (!metadata) {
+		metadata = {
+			basePath: '/',
+			middlewares: [],
+			routes: new Map(),
+		};
+		registry.set(controllerClass, metadata);
+	}
+	return metadata;
+};
+
+export const getRouteMetadata = (controllerClass: Function, handlerName: HandlerName) => {
+	const metadata = getControllerMetadata(controllerClass);
+	let route = metadata.routes.get(handlerName);
+	if (!route) {
+		route = {} as RouteMetadata;
+		metadata.routes.set(handlerName, route);
```

Comment:
Doesn't this break the type safety, if we can't rely anymore that the objects in the route metadata conform to that type? Should we instead validate in the `activateController` that the metadata has been defined correctly and throw if not?

Reply:
> Doesn't this break the type safety

it _kinda_ does. but If someone creates a new controller class that does not use the decorators correctly, the application will already fail to start, so I'm not that worried about it. But I can also add some additional validation in `activateController`, to add more explicit errors.

Reply:
adding the validation seems pretty straight forward, but adding tests for it require resetting the registry, which is making the code a bit more complicated. I'd like to try doing this in another PR.